### PR TITLE
GEODE-9269: Make the lock holding has the same order.

### PR DESCRIPTION
**Pull request by pivotal-eshu**
* When handling TXLockService, always trying to acquire destroy read lock
    before getting the synchronized batchLocks.
  * If destroy read lock is not acuried, thread will wait until the grantor
    is destoryed.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Review by DonalEvans — CHANGES_REQUESTED** [review]
A few typos and test cleanup suggestions. Also, the order in which we take the lock and synchronize is the same as the order in which we release the lock and exit the synchronized block. I'm wondering if it might be better to exit the synchronized block first, then release the lock.

File: geode-core/src/test/java/org/apache/geode/distributed/internal/locks/DLockGrantorTest.java [hunk 1]
```diff
@@ -18,28 +18,40 @@
...
+  private final DLockRequestProcessor.DLockRequestMessage request = mock(
+      DLockRequestProcessor.DLockRequestMessage.class);
+  private final InternalDistributedMember owner = mock(InternalDistributedMember.class);
 
   @Before
   public void setup() {
     dLockService = mock(DLockService.class, RETURNS_DEEP_STUBS);
     DistributionManager distributionManager = mock(DistributionManager.class);
     when(dLockService.getDistributionManager()).thenReturn(distributionManager);
+    when(dLockService.getDLockLessorDepartureHandler())
```
[the remark this answers is not in the document]
Comment (pivotal-eshu) [reply geode-core/src/test/java/org/apache/geode/distributed/internal/locks/DLockGrantorTest.java:53]: I thought the RETURNS_DEEP_STUBS should do the trick, but the test runs failed:
https://concourse.apachegeode-ci.info/teams/main/pipelines/apache-develop-pr/jobs/StressNewTestOpenJDK11/builds/1902

File: geode-core/src/main/java/org/apache/geode/distributed/internal/locks/DLockGrantor.java [hunk 2]
```diff
@@ -639,21 +649,23 @@ public void updateLockBatch(Object batchId, DLockBatch newBatch) throws Interrup
...
+
+      synchronized (batchLocks) {
+        try {
+          checkDestroyed();
+          final DLockBatch oldBatch = batchLocks.get(batchId);
+          if (oldBatch != null) {
+            batchLocks.put(batchId, newBatch);
+          }
+        } finally {
+          releaseDestroyReadLock();
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/locks/DLockGrantor.java:663]: this should be done after the sync is released

File: geode-core/src/main/java/org/apache/geode/distributed/internal/locks/DLockGrantor.java [hunk 3]
```diff
@@ -602,18 +609,20 @@ public DLockBatch getLockBatch(Object batchId) throws InterruptedException {
...
-        releaseDestroyReadLock();
+
+    waitWhileInitializing();
+    if (acquireDestroyReadLock(0)) {
+      synchronized (batchLocks) {
+        try {
+          checkDestroyed();
+          ret = batchLocks.get(batchId);
+        } finally {
+          releaseDestroyReadLock();
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/locks/DLockGrantor.java:620]: this should be done after the sync is released

File: geode-core/src/main/java/org/apache/geode/distributed/internal/locks/DLockGrantor.java [hunk 4]
```diff
@@ -483,46 +484,51 @@ void handleLockBatch(DLockRequestMessage request) throws InterruptedException {
...
+          if (isTraceEnabled_DLS) {
+            logger.trace(LogMarker.DLS_VERBOSE, "[DLockGrantor.handleLockBatch] granting {}",
+                batch.getBatchId());
+          }
+          batchLocks.put(batch.getBatchId(), batch);
+          request.respondWithGrant(Long.MAX_VALUE);
+        } catch (CommitConflictException ex) {
+          request.respondWithTryLockFailed(ex.getMessage());
+        } finally {
+          releaseDestroyReadLock();
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/locks/DLockGrantor.java:518]: this should be done after the sync is released

File: geode-core/src/main/java/org/apache/geode/distributed/internal/locks/DLockGrantor.java [hunk 5]
```diff
@@ -674,24 +687,30 @@ public void releaseLockBatch(Object batchId, InternalDistributedMember owner)
...
+    if (acquireDestroyReadLock(0)) {
+      synchronized (batchLocks) {
+        try {
+          checkDestroyed();
+          DLockBatch batch = batchLocks.remove(batchId);
+          if (batch != null) {
+            releaseReservation(batch);
+          }
+        } finally {
+          releaseDestroyReadLock();
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/locks/DLockGrantor.java:700]: this should be done after the sync is released

**Review by dschneider-pivotal — CHANGES_REQUESTED** [review]
Since you are changing the order we do the readLock and the sync, you should also change the order they are released. Right after you have acquired the read lock you should have the try/finally that has the finally release that read lock. Inside that try block you would have the sync block. Doing it this way makes more sense and allows the code to be refactored (for example you could extra the sync blocks into their own methods). So unless you have a good reason for not have the releases match the acquires, I think you should change them to match.

File: geode-core/src/test/java/org/apache/geode/distributed/internal/locks/DLockGrantorTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (DonalEvans) [reply geode-core/src/test/java/org/apache/geode/distributed/internal/locks/DLockGrantorTest.java:53]: Oh, weird. That must be something to do with how StressNewTest runs things in parallel.

**Review by DonalEvans — APPROVED** [review]

**Review by dschneider-pivotal — APPROVED** [review]

**Pull request merged**